lib/web/apiserver.go

```diff
@@ -2688,6 +2692,60 @@ func (h *Handler) WithClusterAuth(fn ClusterHandler) httprouter.Handle {
 	})
 }
 
+// ProvisionTokenHandler is a authenticated handler that is called for some existing Token
+type ProvisionTokenAuthedHandler func(w http.ResponseWriter, r *http.Request, p httprouter.Params, site reversetunnel.RemoteSite) (interface{}, error)
+
+// WithProvisionTokenAuth ensures that request is authenticated with a provision token.
+// Provision tokens, when used like this are invalidated as soon as used.
+// Doesn't matter if the underlying response was a success or an error.
+func (h *Handler) WithProvisionTokenAuth(fn ProvisionTokenAuthedHandler) httprouter.Handle {
+	return httplib.MakeHandler(func(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
+		ctx := r.Context()
+		logger := h.log.WithField("request", fmt.Sprintf("%v %v", r.Method, r.URL.Path))
+
+		creds, err := roundtrip.ParseAuthHeaders(r)
+		if err != nil {
+			logger.WithError(err).Warn("No auth headers.")
+			return nil, trace.AccessDenied("need auth")
+		}
+
+		if err := h.consumeTokenForAPICall(ctx, creds.Password); err != nil {
+			h.log.WithError(err).Warn("Failed to authenticate.")
+			return nil, trace.AccessDenied("need auth")
+		}
+
+		clusterName := p.ByName("site")
+		if clusterName == currentSiteShortcut {
+			res, err := h.GetProxyClient().GetClusterName()
+			if err != nil {
+				h.log.WithError(err).Warn("Failed to query cluster name.")
+				return nil, trace.Wrap(err)
+			}
+			clusterName = res.GetClusterName()
+		}
+
+		site, err := h.cfg.Proxy.GetSite(clusterName)
```

Comment:
I'm pretty sure web handler already has access to current cluster name, via `h.auth.clusterName`. Can just use that?

```suggestion
		site, err := h.cfg.Proxy.GetSite(h.auth.clusterName)
```